components/dashboard/src/components/UsageBasedBillingConfig.tsx

```diff
@@ -16,27 +17,157 @@ import DropDown from "../components/DropDown";
+    const [pendingStripeSubscription, setPendingStripeSubscription] = useState<PendingStripeSubscription | undefined>();
+    const [billingError, setBillingError] = useState<string | undefined>();
+
+    const localStorageKey = `pendingStripeSubscriptionFor${subject?.id || ""}`;
+
+    useEffect(() => {
+        if (!subject) {
+            return;
+        }
+        (async () => {
+            setStripeSubscriptionId(undefined);
+            setIsLoading(true);
+            try {
+                const subscriptionId = await getGitpodService().server.findStripeSubscriptionId(
+                    AttributionId.render(attributionId),
+                );
+                setStripeSubscriptionId(subscriptionId);
+            } catch (error) {
+                console.error(error);
+            } finally {
+                setIsLoading(false);
+            }
+        })();
+    }, [subject]);
+
+    useEffect(() => {
+        if (!subject || !stripeSubscriptionId) {
+            return;
+        }
+        (async () => {
+            const [portalUrl, spendingLimit] = await Promise.all([
+                getGitpodService().server.getStripePortalUrl(AttributionId.render(attributionId)),
+                getGitpodService().server.getUsageLimit(AttributionId.render(attributionId)),
+            ]);
+            setStripePortalUrl(portalUrl);
+            setUsageLimit(spendingLimit);
+        })();
+    }, [subject, stripeSubscriptionId]);
+
+    useEffect(() => {
+        if (!subject) {
+            return;
+        }
+        const params = new URLSearchParams(location.search);
+        if (!params.get("setup_intent") || params.get("redirect_status") !== "succeeded") {
+            return;
+        }
+        (async () => {
+            const setupIntentId = params.get("setup_intent")!;
+            window.history.replaceState({}, "", location.pathname);
+            const pendingSubscription = { pendingSince: Date.now() };
+            setPendingStripeSubscription(pendingSubscription);
+            window.localStorage.setItem(localStorageKey, JSON.stringify(pendingSubscription));
+            try {
+                await getGitpodService().server.subscribeToStripe(AttributionId.render(attributionId), setupIntentId);
+            } catch (error) {
+                console.error("Could not subscribe subject to Stripe", error);
+                window.localStorage.removeItem(localStorageKey);
+                clearTimeout(pollStripeSubscriptionTimeout!);
+                setPendingStripeSubscription(undefined);
+                setBillingError(`Could not subscribe subject to Stripe. ${error?.message || String(error)}`);
+            }
+        })();
+    }, [location.search, subject]);
+
+    useEffect(() => {
+        setPendingStripeSubscription(undefined);
+        if (!subject) {
+            return;
+        }
+        try {
+            const pendingStripeSubscription = window.localStorage.getItem(localStorageKey);
```

Comment:
Why is pendingStripeSubscription a global and not local state?

Reply:
I see it used to be like this before this change, just wonder what the reason is.

Reply:
@svenefftinge Can you do a local state that survives navigations and page reloads? 👀

Reply:
Of course not. Why does it need to survive that?

Reply:
We've seen that users like to reload billing/subscription pages often.

The currently-persisted "pending subscription" allows the following UX:
1. Enter credit card details
2. They go through, we're subscribing you to Stripe
3. You reload the page -- we're still subscribing you to Stripe and waiting for confirmation

Without the persistence, it changes to:
1. Enter credit card details
2. They go through, we're subscribing you to Stripe
3. You reload the page -- it looks like it didn't work, so maybe you immediately try to add your credit card again